src/storage/src/hummock/compactor/iterator.rs

```diff
@@ -426,6 +399,91 @@ impl ConcatSstableIterator {
         }
         Ok(())
     }
+
+    pub fn filter_block_metas(
+        block_metas: &Vec<BlockMeta>,
+        seek_key: Option<FullKey<&[u8]>>,
+        existing_table_ids: &HashSet<u32>,
+        key_range: &KeyRange,
+    ) -> (usize, usize) {
+        if block_metas.is_empty() {
+            return (usize::MAX, 0);
+        }
+
+        let mut start_index = match seek_key {
+            None => 0,
+            Some(seek_key) => {
+                // start_index points to the greatest block whose smallest_key <= seek_key.
+                block_metas
+                    .partition_point(|block| {
+                        seek_key.cmp(&FullKey::decode(&block.smallest_key)) != Ordering::Less
+                    })
+                    .saturating_sub(1)
+            }
+        };
+
+        let mut end_index = if key_range.right.is_empty() {
+            block_metas.len()
+        } else {
+            block_metas.partition_point(|block| {
+                KeyComparator::compare_encoded_full_key(&block.smallest_key, &key_range.right)
+                    != Ordering::Greater
+            })
+        };
+
+        if end_index == 0 {
+            return (usize::MAX, 0);
+        }
+
+        end_index = end_index.saturating_sub(1);
+
+        // skip blocks that are not in existing_table_ids
+        while start_index <= end_index {
+            let start_block_table_id = block_metas[start_index].table_id().table_id();
+            if existing_table_ids.contains(&start_block_table_id) {
+                break;
+            }
+
+            // skip this table_id
+            let old_start_index = start_index;
+            let block_metas_to_search = &block_metas[start_index..=end_index];
+
+            start_index += block_metas_to_search.partition_point(|block_meta| {
+                block_meta.table_id().table_id() == start_block_table_id
+            });
+
+            if old_start_index == start_index {
+                // no more blocks with the same table_id
+                break;
+            }
+        }
+
+        if end_index != 0 {
+            while start_index <= end_index {
+                let end_block_table_id = block_metas[end_index].table_id().table_id();
+                if existing_table_ids.contains(&end_block_table_id) {
+                    break;
+                }
+
+                let old_end_index = end_index;
+                let block_metas_to_search = &block_metas[start_index..=end_index];
+
+                end_index = start_index
+                    + block_metas_to_search
+                        .partition_point(|block_meta| {
+                            block_meta.table_id().table_id() < end_block_table_id
```

Comment:
Will the logic here be simpler with two `while`, one for start_index, one for end_index？We can handle both the key range and table id filtering logic together in the while loop.

Reply:
I'm sorry, I didn't fully understand what you meant.

1.The key range filtering and requiring a while loop, which is a one-time operation
2. The table_id filtering is already done with two while loops.

I've refactored the code to make it more readable.